injection/sharedmain/main.go

```diff
@@ -143,102 +166,180 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
+		}
 
-	// Start all of the controllers.
-	logger.Info("Starting controllers...")
-	go controller.StartAll(ctx.Done(), controllers...)
+		// Start all of the controllers.
+		logger.Info("Starting controllers...")
+		go controller.StartAll(ctx.Done(), controllers...)
+
+		profilingServer := profiling.NewServer(profilingHandler)
+
+		eg, egCtx := errgroup.WithContext(ctx)
+		eg.Go(profilingServer.ListenAndServe)
+
+		// If we have one or more admission controllers, then start the webhook
+		// and pass them in.
+		if len(webhooks) > 0 {
+			// Register webhook metrics
+			webhook.RegisterMetrics()
+
+			// possible bug? egCtx
+			wh, err := webhook.New(ctx, webhooks)
+			if err != nil {
+				logger.Fatalw("Failed to create admission controller", zap.Error(err))
+			}
+			eg.Go(func() error {
+				return wh.Run(ctx.Done())
+			})
+		}
 
-	profilingServer := profiling.NewServer(profilingHandler)
+		// This will block until either a signal arrives or one of the grouped functions
+		// returns an error.
+		<-egCtx.Done()
 
-	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(profilingServer.ListenAndServe)
+		profilingServer.Shutdown(context.Background())
+		// Don't forward ErrServerClosed as that indicates we're already shutting down.
+		if err := eg.Wait(); err != nil && err != http.ErrServerClosed {
+			logger.Errorw("Error while running server", zap.Error(err))
+		}
+	}
 
-	// If we have one or more admission controllers, then start the webhook
-	// and pass them in.
-	if len(webhooks) > 0 {
-		// Register webhook metrics
-		webhook.RegisterMetrics()
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		// Create event broadcaster
+		logger.Debug("Creating event broadcaster")
+		eventBroadcaster := record.NewBroadcaster()
+		watches := []watch.Interface{
+			eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof),
+			eventBroadcaster.StartRecordingToSink(
+				&typedcorev1.EventSinkImpl{Interface: kubeclient.Get(ctx).CoreV1().Events("")}),
+		}
+		recorder = eventBroadcaster.NewRecorder(
+			// todo: what agent name?
+			scheme.Scheme, corev1.EventSource{Component: "fake-agent-name"})
+		go func() {
+			<-ctx.Done()
+			for _, w := range watches {
+				w.Stop()
+			}
+		}()
+	}
 
-		wh, err := webhook.New(ctx, webhooks)
+	// Set up leader election config
+	leaderElectionConfig, err := GetLeaderElectionConfig(ctx)
+	if err != nil {
+		log.Fatalf("Error loading leader election configuration: %v", err)
+	}
+	leConfig := leaderElectionConfig.GetComponentConfig(component)
+
+	if !leConfig.LeaderElect {
+		log.Printf("%v will not run in leader-elected mode", component)
+		run(ctx)
+		logger.Fatal("unreachable")
+	} else {
+		// Create a unique identifier so that two controllers on the same host don't
+		// race.
+		id, err := os.Hostname()
+		if err != nil {
+			logger.Fatalw("Failed to get hostname for leader election", zap.Error(err))
+		}
+		id += "_" + string(uuid.NewUUID())
```

Comment:
Can we make this a helper in the leaderelection package?  Seems like you use it in serving as well.

Reply:
yes